### ISSUE TITLE
Injecting into system apps on android 10 fails due to "execute-only (no-read)" permissions.

### ISSUE THREAD

**Issue by Alien-AV** [frida/frida#1043]
This happens due to change of memory mapping of system apps, described [in Android Developer documentation](https://developer.android.com/about/versions/10/behavior-changes-all#xom-binaries). As a workaround it suggests using `mprotect()` to change the permissions, however I don't know where best to invoke it.

Also I don't know why/how Frida is looking at the memory without changing permissions to rwx first, even before Android 10.

Crash below due to injection into Google Maps application (installed as system).
```
Build fingerprint: 'google/crosshatch/crosshatch:10/QP1A.191005.007/5878874:user/release-keys'
Revision: 'MP1.0'
ABI: 'arm64'
Timestamp: 2019-10-10 18:03:45+0300
pid: 29415, tid: 4966, name: droid.apps.maps  >>> com.google.android.apps.maps <<<
uid: 10114
signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0x7c241c688b
Cause: execute-only (no-read) memory access error; likely due to data in .text.
    x0  0000007b871f0700  x1  0000000000000000  x2  ffffffffffffffe0  x3  0000007b871f0880
    x4  0000007b871f08e0  x5  0000000000000004  x6  0000007b871f05a0  x7  0000000000000000
    x8  0000000000000000  x9  00000000ffffffff  x10 0000007b871f0700  x11 0000007b871f0e40
    x12 0000000000000000  x13 0000007aec297160  x14 0000007aec2aa218  x15 0000007b8739ddc0
    x16 0000007aeae9c600  x17 0000007c24165340  x18 0000007b11ee2000  x19 0000007b1356f044
    x20 0000007b1356f048  x21 0000007b871f04e0  x22 0000007c241c6888  x23 0000000000000001
    x24 0000007c241c6888  x25 00000000000000f8  x26 0000007aec2b7170  x27 0000007c241c6888
    x28 0000000000000010  x29 0000007b1356edf0
    sp  0000007b1356edc0  lr  0000007ae9e722b0  pc  0000007ae9e722d4

backtrace:
      #00 pc 00000000002452d4  /data/local/tmp/re.frida.server/frida-agent-64.so
```

**Comment by oleavr** [comment]
It's likely the Android linker detection code or Interceptor (both in Gum). For the latter we could flip to `rwx` before we start, and keep it like that. (We currently flip to `rwx` before making the modifications, so we'd just have to move that earlier.) But yeah if you build Frida with symbols – by removing `--strip` from `config.mk` – you'll see exactly where in the backtrace.

**Comment by esanfelix** [comment]
A similar crash to the one above happens when launching `frida-server` too, which causes `system_server` to crash and as a result the Android framework soft-resets.

I built frida-server wit symbols as described by @oleavr above, and got crash logs for both the `system_server` case as well as attaching to a `cat` process. In both cases the backtrace only contains one item:

```
pid: 6006, tid: 19410, name: system_server  >>> system_server <<<
uid: 1000
signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0x792e53788b
Cause: execute-only (no-read) memory access error; likely due to data in .text.
    x0  00000078459a4700  x1  0000000000000000  x2  ffffffffffffffe0  x3  00000078459a4880
    x4  00000078459a48e0  x5  0000000000000004  x6  00000078459a45a0  x7  0000000000000000
    x8  0000000000000000  x9  00000000ffffffff  x10 00000078459a4700  x11 00000078459a4e40
    x12 0000000000000000  x13 0000007845724160  x14 0000007845987218  x15 0000007845a7ff00
    x16 00000077a2649600  x17 000000792e4d6340  x18 00000077a2f4c000  x19 00000077c96f6044
    x20 00000077c96f6048  x21 00000078459a44e0  x22 000000792e537888  x23 0000000000000001
    x24 000000792e537888  x25 00000000000000f8  x26 0000007845741fa0  x27 000000792e537888
    x28 0000000000000010  x29 00000077c96f5df0
    sp  00000077c96f5dc0  lr  00000077a1643378  pc  00000077a164339c

backtrace:
      #00 pc 000000000027939c  /data/local/tmp/re.frida.server/frida-agent-64.so (AArch64_getInstruction+104)

stack:
         0000007c61118d40  2e6d6f632f786570
         0000007c61118d48  2e64696f72646e61

```

The backtrace for `cat` is very similar, which I'm afraid doesn't provide a lot of additional information.

I don't have a lot of time to put into this, but I'll give it a try to see if I can get more information. One thought is to just `mprotect` anything that goes through this function to see if that's the only problem we are facing. 

After that we may discuss further how to best approach a fix.

**Comment by esanfelix** [comment]
@oleavr mentioned that capstone is always built without debug info, and instructed me to remove these lines: https://github.com/frida/frida/blob/fa888900710f3869582c5b4b66153f1578ec4dfb/releng/config.site.in#L17-L20

After doing so, I no longer get the SIGSEGV but an `abort` with a full stack trace. The first `abort` occurs in https://github.com/frida/frida-java-bridge/blob/master/lib/android.js#L507, which I've patched as follows:

```diff
diff --git a/lib/android.js b/lib/android.js
index 28c540d..55b5436 100644
--- a/lib/android.js
+++ b/lib/android.js
@@ -501,6 +501,7 @@ function tryDetectInstrumentationOffset () {
     return null;
   }
 
+  Memory.protect(cur, 0x1000, 'rwx');
   const tryParse = instrumentationOffsetParsers[Process.arch];
 
   for (let i = 0; i !== 20; i++) {
```

Now I get to the next abort, which is again inside frida-java-bridge, in https://github.com/frida/frida-java-bridge/blob/master/lib/android.js#L2253 . My  understanding is that this is caused by Interceptor trying to parse the code, so most likely the best approach is to change it in Interceptor itself.

**Comment by oleavr** [comment]
@esanfelix Thanks for digging into this! Yes, we will need something along those lines. The drawback is that the above solution will introduce one additional system call in overhead per function, so we don't get the benefits of our current transaction support where we only flip page protection once per page when hooking a group of functions on the same page. I think the cleanest solution is to introduce a cache local to the transaction, so we can quickly determine if we have already flipped a given page. Ideally this would also remember the current protection, so we can restore it to the exact same one afterwards. (Which we currently don't – we currently blindly flip back to RX, which isn't good at all.)

**Comment by ch0rus** [comment]
@esanfelix Cloud you possibly share a compiled version with you propsed patch?

**Comment by oleavr** [comment]
Fix is out in Frida 12.7.14:
https://twitter.com/fridadotre/status/1188585509737652224

**Comment by SajjadPourali** [comment]
@oleavr: it seems the problem still existing on the thread backtrace.
<img width="1085" alt="mprotect" src="https://user-images.githubusercontent.com/20374762/107625231-43d3f280-6c71-11eb-96f2-6360dd11c2f7.png">
<img width="893" alt="Screen Shot 2021-02-10 at 15 51 05 copy" src="https://user-images.githubusercontent.com/20374762/107625242-4898a680-6c71-11eb-8a59-bfe6cca7d5bb.png">
Frida 14.2.12
